### PR TITLE
[tool/fpv] add fpv to dvsim.py

### PR DESCRIPTION
**Pull request by cindychip**
Support adding formal to dvsim.py flow
Add full support for jaspergold and partial support for VC Formal
TODOs:

- [x] work with Cadence to see if we can publish the results

- [x] Run performance analysis to see why when running all modules, the machine hangs

Here is an example of the [FPV dashboard](https://reports.opentitan.org/hw/top_earlgrey/fpv/summary.html).

To test run, you can go to `hw/top_earlgrey/fpv/`, use command:
`../../../util/dvsim/dvsim.py top_earlgrey_fpv_cfgs.hjson -t jaspergold -mp 1`

Signed-off-by: Cindy Chen <chencindy@google.com>

### CODE REVIEW

File: util/dvsim/FpvCfg.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+                    os.remove(cov_data)
+                except IOError as err:
+                    log.error("%s", err)
+                    results_str += str(err)
+
+        if table != [header]:
+            self.results_md = results_str + tabulate(
+                table, headers="firstrow", tablefmt="pipe", colalign=colalign)
+        else:
+            self.results_md = results_str
```
Comment (cindychip) [inline util/dvsim/FpvCfg.py:143]: The performance issue is due to the "tabulate" function here. When modules like `top_earlgrey` has thousands of properties, each property is an item in the tabulates. This blows up the calculation. I am going to only print out the summary.

File: hw/formal/tools/jaspergold/fpv.tcl [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+#-------------------------------------------------------------------------
+# specify clock(s) and reset(s)
+#-------------------------------------------------------------------------
+
+# select primary clock and reset condition (use ! for active-low reset)
+# note: -both_edges is needed below because the TL-UL protocol checker
+# tlul_assert.sv operates on the negedge clock
+
+if {$env(FPV_TOP) == "rv_dm"} {
```
Comment (sriyerg) [inline hw/formal/tools/jaspergold/fpv.tcl:38]: Ideally, this file should be agnostic to $FPV_TOP.

File: hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+// SPDX-License-Identifier: Apache-2.0
+{
+
+  // This is the master cfg hjson for FPV. It imports ALL individual FPV
+  // dedicated cfgs of the IPs. This enables to run them all as a regression in one shot.
+  name: top_earlgrey_dedicated_ip_batch_fpv
+
+  import_cfgs: [// common server configuration for results upload
+                "{proj_root}/hw/data/common_project_cfg.hjson"
+                "{proj_root}/hw/formal/data/dedicated_fpv_cfgs.hjson"]
```
Comment (sriyerg) [inline hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson:12]: A master cfg list should ideally not be imported in other cfgs. It needs to be the entry point to the tool as input. What is the intent of organizing it this way?

Comment (cindychip) [reply hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson:12]: I see, I will go back the read the `testplan` readme. Here I wanted to distinguish two sets of FPV testbenches:
1. `{proj_root}/hw/data/common_project_cfg.hjson` includes DUTs that partially verified by FPV: only includes bind file with TLUL assertions and internal assertions. (There is no need to collect coverage on these modules)
2. `{proj_root}/hw/formal/data/dedicated_fpv_cfgs.hjson` includes DUTs that fully verified by FPV. (They need to collect coverage)
3. `hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson` runs both 1) and 2) modules as a weekly regression results.

File: hw/formal/tools/jaspergold/fpv.tcl (hunk 2 above)
Comment (sriyerg) [reply hw/formal/tools/jaspergold/fpv.tcl:38]: Sure, no need to address this in this PR.

File: hw/formal/data/common_fpv_cfg.hjson [hunk 4]
```diff
@@ -0,0 +1,36 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+{
+  flow:             fpv
+  flow_makefile:    "{proj_root}/hw/formal/data/fpv.mk"
+
+  import_cfgs:      [// common server configuration for results upload
+                     "{proj_root}/hw/data/common_project_cfg.hjson",
+                     "{proj_root}/hw/formal/data/jaspergold.hjson"]
```
Comment (sriyerg) [inline hw/formal/data/common_fpv_cfg.hjson:10]: "Common" cfg should ideally not contain tool-specific settings.

Comment (sriyerg) [reply hw/formal/data/common_fpv_cfg.hjson:10]: Is this better?

```suggestion
                     "{proj_root}/hw/formal/data/{tool}/{tool}.hjson"]
```

File: hw/formal/data/jaspergold.hjson [hunk 5]
```diff
@@ -0,0 +1,15 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+{
+  build_cmd: "jg -batch fpv.tcl"
```
[the remark this answers is not in the document]
Comment (sriyerg) [reply hw/formal/data/jaspergold.hjson:5]: Should this be referenced from `{tool_srcs_dir}`?

File: hw/formal/fpv.tcl [hunk 6]
```diff
@@ -167,13 +167,14 @@ set_proofgrid_per_engine_max_local_jobs 2
 #-------------------------------------------------------------------------
```
Comment (sriyerg) [inline hw/formal/fpv.tcl:167]: I assume this is not used by DVSim? I think it would be better to move this to a separate commit. Also, this should probably be renamed to `jaspergold_fpv.tcl` to distinguish from VCF one.

File: hw/formal/tools/vcformal/fpv_vcf.tcl [hunk 7]
```diff
@@ -0,0 +1,163 @@
+# Copyright lowRISC contributors.
```
Comment (sriyerg) [inline hw/formal/tools/vcformal/fpv_vcf.tcl:1]: I think renaming this to `fpv.tcl` should enable you move things to `common_fpv_cfg.hjson`.

File: hw/formal/data/common_fpv_cfg.hjson (hunk 4 above)
Comment (cindychip) [reply hw/formal/data/common_fpv_cfg.hjson:10]: Thanks Sri, yes that works better.

File: hw/formal/fpv.tcl (hunk 6 above)
Comment (cindychip) [reply hw/formal/fpv.tcl]: Yeah that is right, thanks for the reminder. I will submit a separate PR for that one.

File: hw/formal/data/dedicated_fpv_cfgs.hjson [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+// This hjson file list all the IPs that are fully verified by FPV
+{
+  use_cfgs: [
+             {
+               name: prim_arbiter_ppc_fpv
+               fusesoc_core: lowrisc:fpv:prim_arbiter_ppc_fpv
+               import_cfgs: ["{proj_root}/hw/formal/data/common_fpv_cfg.hjson"]
+               rel_path: "hw/ip/prim/fpv"
```
Comment (sriyerg) [inline hw/formal/data/dedicated_fpv_cfgs.hjson:11]: Try adding `tool: jaspergold` so that `-t jaspergold` doesn't have to be specified on the command line.

File: hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson (hunk 3 above)
Comment (sriyerg) [reply hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson:12]: It is somehow working and that is good to see. Let's leave it at that and fix any tool issues we might run into later.

File: hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  use_cfgs:    []
```
Comment (sriyerg) [inline hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson:13]: This should not be needed.

File: util/dvsim/FpvCfg.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+                        self.result["pass_rate"],
+                        self.result["cov_rate"]])
+                os.remove(result_data)
+            except IOError as err:
+                log.error("%s", err)
+                results_str += str(err)
+
+            if (self.cov):
+                cov_data = Path(
+                    subst_wildcards(self.build_dir, {"build_mode": mode.name}) + "/cov.csv")
```
Comment (sriyerg) [inline util/dvsim/FpvCfg.py:155]: This file will not exist if `--dry-run or -n` switch is passed, and an error will be thrown.

File: hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson (hunk 9 above)
Comment (cindychip) [reply hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson:13]: I guess this would be the first issue we encounter - If I take it away, I will get this error.. 
`ERROR: [FlowCfg] Key "use_cfgs" encountered in a non-master cfg file list "/usr/local/google/home/chencindy/opentitan/hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson"`
I can work on fixing this in a separate PR. If the solution is not desired, I can revert back the code and not import master cfgs.

File: hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson (hunk 3 above)
Comment (cindychip) [reply hw/top_earlgrey/fpv/top_earlgrey_dedicated_fpv_cfgs.hjson:12]: Sounds good, If happened, I will work on the fix related to this.

File: util/dvsim/FpvCfg.py (hunk 10 above)
Comment (cindychip) [reply util/dvsim/FpvCfg.py:155]: Thanks for pointing this out. I fixed them from throwing error to warning.

File: hw/formal/data/dedicated_fpv_cfgs.hjson (hunk 8 above)
Comment (cindychip) [reply hw/formal/data/dedicated_fpv_cfgs.hjson:11]: That is a good idea, but seems like `OneShotCfg.py` does not allow the argument `-t` to be empty (https://github.com/lowRISC/opentitan/blob/master/util/dvsim/OneShotCfg.py#L26). I will work with Michael to see if he is also in flavor of removing this constraint and add a default tool to each flow.

File: util/dvsim/FpvCfg.py [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+                str(cfg.result["unreachable"]) + " W ",
+                cfg.result["pass_rate"],
+                cfg.result["cov_rate"]])
+        if len(email_table) > 1:
+            self.email_summary_md = results_str + tabulate(
+                email_table, headers="firstrow", tablefmt="pipe", colalign=colalign)
+
+        return self.results_summary_md
+
+    def _gen_results(self):
```
Comment (msfschaffner) [inline util/dvsim/FpvCfg.py:77]: I think it would be helpful if we aggregated and presented tool warnings and errors here as well. that way we will know whether an FPV pass completed correctly or not. e.g., sometimes runs are not successful because of syntax errors or similar... 

I think you already have an `["errors"]` field in the parsed out hjson data structure that is populated if the logfiles cannot be read. we could copy over the regex search mechanism from lint and adapt the patterns for FPV to cover more tool warnings and errors:
https://github.com/lowRISC/opentitan/blob/b33ffdaa35c399080d95042fd79789d276e4384b/hw/lint/tools/ascentlint/parse-lint-report.py#L40-L62

we just have to be careful with the messages themselves, since we may not be able to print them out publicly. but reporting the error/warning count (as e.g. done for synthesis) should be ok.

**Comment by cindychip** [comment]
Thank you and Sri very much for reviewing this long PR and providing
feedback. Sorry I should have split it into smaller ones for the ease of
review. But this is my first time drafting tool related PRs, and I was not
very sure about the quality until I could see FPV runs on it.
But now I can definitely split it into smaller PRs. I will address your
comments in the splitted PRs, and then close this one.

For the general question: Yes, I plan to support VC Formal. I will refer to
your Linting solution on supporting both of them. thank you for the
notification.


  Cindy Chen

**Comment by cindychip** [comment]
As michael suggested, this PR will be replaced by a few more PRs for the ease of review.
